### PR TITLE
Interpolate pos encode vitmae

### LINKED ISSUE #30579: Community contribution: enable dynamic resolution input for more vision models.
### Feature request

Some of our models interpolate its positional embeddings, enabling pretrained checkpoints to be used on different input resolutions. For example, [here in ViT](https://github.com/huggingface/transformers/blob/75bbfd5b2237b7e35a9265731ecf63022579e7e2/src/transformers/models/vit/modeling_vit.py#L79). 

- [ ] [altclip](https://github.com/huggingface/transformers/blob/main/src/transformers/models/altclip/modeling_altclip.py)
- [ ] [beit](https://github.com/huggingface/transformers/blob/main/src/transformers/models/beit/modeling_beit.py)
- [ ] [blip](https://github.com/huggingface/transformers/blob/main/src/transformers/models/blip/modeling_blip.py)
- [ ] [blip_2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/blip_2/modeling_blip_2.py)
- [ ] [bridgetower](https://github.com/huggingface/transformers/blob/main/src/transformers/models/bridgetower/modeling_bridgetower.py)
- [ ] [chinese_clip](https://github.com/huggingface/transformers/blob/main/src/transformers/models/chinese_clip/modeling_chinese_clip.py)
- [ ] [clip](https://github.com/huggingface/transformers/blob/main/src/transformers/models/clip/modeling_clip.py)
- [ ] [conditional_detr](https://github.com/huggingface/transformers/blob/main/src/transformers/models/conditional_detr/modeling_conditional_detr.py)
- [ ] [data2vec](https://github.com/huggingface/transformers/blob/main/src/transformers/models/data2vec/modeling_data2vec_vision.py)
- [ ] [deformable_detr](https://github.com/huggingface/transformers/blob/main/src/transformers/models/deformable_detr/modeling_deformable_detr.py)
- [ ] [deit](https://github.com/huggingface/transformers/blob/main/src/transformers/models/deit/modeling_deit.py)
- [ ] [detr](https://github.com/huggingface/transformers/blob/main/src/transformers/models/detr/modeling_detr.py)
- [ ] [donut](https://github.com/huggingface/transformers/blob/main/src/transformers/models/donut/modeling_donut_swin.py)
- [ ] [git](https://github.com/huggingface/transformers/blob/main/src/transformers/models/git/modeling_git.py)
- [ ] [grounding_dino](https://github.com/huggingface/transformers/blob/main/src/transformers/models/grounding_dino/modeling_grounding_dino.py)
- [ ] [idefics2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/idefics2/modeling_idefics2.py)
- [ ] [kosmos2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/kosmos2/modeling_kosmos2.py)
- [ ] [maskformer swin](https://github.com/huggingface/transformers/blob/main/src/transformers/models/maskformer/modeling_maskformer_swin.py)
- [ ] [owlv2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/owlv2/modeling_owlv2.py)
- [ ] [owlvit](https://github.com/huggingface/transformers/blob/main/src/transformers/models/owlvit/modeling_owlvit.py)
- [ ] [perceiver](https://github.com/huggingface/transformers/blob/main/src/transformers/models/perceiver/modeling_perceiver.py)
- [ ] [siglip](https://github.com/huggingface/transformers/blob/main/src/transformers/models/siglip/modeling_siglip.py)
- [ ] [swin](https://github.com/huggingface/transformers/blob/main/src/transformers/models/swin/modeling_swin.py)
- [ ] [swin2sr](https://github.com/huggingface/transformers/blob/main/src/transformers/models/swin2sr/modeling_swin2sr.py)
- [ ] [swinv2](https://github.com/huggingface/transformers/blob/main/src/transformers/models/swinv2/modeling_swinv2.py)
- [ ] [table_transformer](https://github.com/huggingface/transformers/blob/main/src/transformers/models/table_transformer/modeling_table_transformer.py)
- [ ] [tvp](https://github.com/huggingface/transformers/blob/main/src/transformers/models/tvp/modeling_tvp.py)
- [ ] [vit_mae](https://github.com/huggingface/transformers/blob/main/src/transformers/models/vit_mae/modeling_vit_mae.py)
- [ ] [vivit](https://github.com/huggingface/transformers/blob/main/src/transformers/models/vivit/modeling_vivit.py)

### Motivation

Let's add this to more models, to leverage existing checkpoints for new cases! 

### Your contribution

For anyone who would like to contribute, please comment on the issue, claiming a model you'd like to work on and share a link to the PR. 

Each PR should: 
* Add an `interpolate_pos_encoding` method
* Add a test showing the model can correctly interpolate an input image of a different size

There was a PR opened to add this to CLIP models, which is now inactive, but useful for reference of the changes to make: https://github.com/huggingface/transformers/pull/27457

Once the PR is ready, you can ping me for review 🤗

### PR DESCRIPTION
**Pull request by bhuvanmdev**
# What does this PR do?
Adds interpolation capabilities for the tf and pytorch versions of vit_mae along with the test cases.

Addresses #30579 (vit_mae)

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested.

@amyeroberts

### CODE REVIEW

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 1]
```diff
@@ -892,17 +982,21 @@ def unpatchify(self, patchified_pixel_values):
                 Pixel values.
         """
         patch_size, num_channels = self.config.patch_size, self.config.num_channels
-        num_patches_one_direction = int(patchified_pixel_values.shape[1] ** 0.5)
+        if not (original_img[2] == original_img[3] == 244):
+            num_patches_one_direction = original_img[2] // patch_size
+            num_patches_two_direction = original_img[3] // patch_size
+        else:
+            num_patches_one_direction = num_patches_two_direction = int(patchified_pixel_values.shape[1] ** 0.5)
```
Comment (amyeroberts) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:989]: We can just use the first case, without having the do the conditional

```suggestion
        original_height, original_width = original_image_size
        num_patches_h = original_height // patch_size
        num_patches_w = original_width // patch_size
```

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 2]
```diff
@@ -892,17 +982,21 @@ def unpatchify(self, patchified_pixel_values):
...
         # sanity check
-        if num_patches_one_direction**2 != patchified_pixel_values.shape[1]:
+        if num_patches_one_direction * num_patches_two_direction != patchified_pixel_values.shape[1]:
             raise ValueError("Make sure that the number of patches can be squared")
```
Comment (amyeroberts) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:992]: Message here should be updated - it doesn't need to be square anymore

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 3]
```diff
@@ -892,17 +982,21 @@ def unpatchify(self, patchified_pixel_values):
...
 
         # unpatchify
         batch_size = patchified_pixel_values.shape[0]
         patchified_pixel_values = patchified_pixel_values.reshape(
             batch_size,
             num_patches_one_direction,
```
Comment (amyeroberts) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:998]: let's rename here - it's a bit clearer than "one direction"

```suggestion
            num_patches_h,
```

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 4]
```diff
@@ -892,17 +982,21 @@ def unpatchify(self, patchified_pixel_values):
...
-            num_patches_one_direction,
+            num_patches_two_direction,
```
Comment (amyeroberts) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:999]:
```suggestion
            num_patches_w,
```

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 5]
```diff
@@ -912,11 +1006,11 @@ def unpatchify(self, patchified_pixel_values):
             batch_size,
             num_channels,
             num_patches_one_direction * patch_size,
```
Comment (amyeroberts) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:1008]:
```suggestion
            num_patches_h * patch_size,
```

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 6]
```diff
@@ -912,11 +1006,11 @@ def unpatchify(self, patchified_pixel_values):
...
-            num_patches_one_direction * patch_size,
+            num_patches_two_direction * patch_size,
```
Comment (amyeroberts) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:1009]:
```suggestion
            num_patches_w * patch_size,
```

File: src/transformers/models/vit_mae/modeling_vit_mae.py [hunk 7]
```diff
@@ -892,17 +982,21 @@ def unpatchify(self, patchified_pixel_values):
                 Pixel values.
         """
         patch_size, num_channels = self.config.patch_size, self.config.num_channels
```
Comment (amyeroberts) [inline src/transformers/models/vit_mae/modeling_vit_mae.py:984]:
```suggestion
        patch_size, num_channels = self.config.patch_size, self.config.num_channels
        original_size = original_size if original_size is not None else (self.config.image_size, self.config.image_size)
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! Overall looking good - just some small nits and implementation comments

File: src/transformers/models/vit_mae/modeling_vit_mae.py (hunk 2 above)
Comment (bhuvanmdev) [reply src/transformers/models/vit_mae/modeling_vit_mae.py:986]: True, then what should i update the new message to?
does the below one sound good?
"**The number of patches in the patchified pixel values is {patchified_pixel_values.shape[1]}, does not match the number of patches on original image {num_patches_w}*{num_patches_h}**"

File: tests/models/vit_mae/test_modeling_vit_mae.py [hunk 8]
```diff
@@ -289,18 +289,14 @@ def prepare_img():
 @require_torch
 @require_vision
 class ViTMAEModelIntegrationTest(unittest.TestCase):
-    @cached_property
-    def default_image_processor(self):
-        return ViTImageProcessor.from_pretrained("facebook/vit-mae-base") if is_vision_available() else None
```
[the remark this answers is not in the document]
Comment (bhuvanmdev) [reply tests/models/vit_mae/test_modeling_vit_mae.py:294]: I assumed the `do_resize` flag can be passed only from the `from_pretrained` class method.
Will undo the removal.

**Comment by bhuvanmdev** [comment]
Due to my negligence, I ended up polluting this PR. will create a new PR for this issue. I'm really sorry for the inconvinience :( .

**Comment by bhuvanmdev** [comment]
Sorry I got confused with one of my older branch and ended up polluting the PR. Will clean up the mess.

**Pull request closed without merging**